### PR TITLE
data import, name corrections, adding mouzas

### PR DESCRIPTION
**Pull request by benkyd**
This is a big PR:
- [x] import the 1m dataset
- [x] add name corrections – in progress to check Mymensing (without h)
- [x] add mouzas – not quite everywhere yet, see review comment
- [x] add flooding model
- [x] put benchmark data in LFS
- [x] lint
- [x] document things
- [x] add CI to lint and test
- [x] remove old R scripts
- [x] compute wells distances on demand, not all N^2 immediately (OOM)
- [x] (with Jacek) edit 6ee2407 to remove 1M dataset file change
- [ ] (with Jacek)  deploy model 5 one last time

### CODE REVIEW

**Comment by jacekkopecky** [comment]
this wants to be against the other branch for now please so the commit history is clear about what this PR does

**Comment by benkyd** [comment]
> this wants to be against the other branch for now please so the commit history is clear about what this PR does

The base of the PR is now data-conversion-final-tidyup

File: preprocessing/README.md [hunk 1]
```diff
@@ -3,7 +3,7 @@
 ## Requirements
 
 * Ensure the csv headers are as follows (**Case matters!**):
-  * `Division,District,Upazila,Depth,Arsenic,Union`
+  * `Division,District,Upazila,Mouza,Depth,Arsenic,Union`
```
Comment (jacekkopecky) [inline preprocessing/README.md:6]: this order must be reviewed

File: preprocessing/lib/load-data.js [hunk 2]
```diff
@@ -85,9 +85,20 @@ function extractLocations(records) {
     if (!(r.Union in upazila.unions)) {
       upazila.unions[r.Union] = {
         wells: [],
+        mouzas: {},
         name: r.Union,
       };
     }
+
+    const union = upazila.unions[r.Mouza];
+    if (union !== undefined) {
```
Comment (jacekkopecky) [inline preprocessing/lib/load-data.js:94]: probably remove this condition

File: preprocessing/lib/load-data.js [hunk 3]
```diff
@@ -85,9 +85,20 @@ function extractLocations(records) {
     if (!(r.Union in upazila.unions)) {
       upazila.unions[r.Union] = {
         wells: [],
+        mouzas: {},
         name: r.Union,
       };
     }
+
+    const union = upazila.unions[r.Mouza];
```
Comment (jacekkopecky) [inline preprocessing/lib/load-data.js:93]: r.Union there?

File: preprocessing/lib/load-data.js [hunk 4]
```diff
@@ -116,6 +128,7 @@ function fillArsenicData(divisions, records) {
     district.wells.push(well);
     upazila.wells.push(well);
     union.wells.push(well);
+    mouza?.wells.push(well);
```
Comment (jacekkopecky) [inline preprocessing/lib/load-data.js:131]: opt chaining should not be necessary

File: preprocessing/models/model5-estimator.js [hunk 5]
```diff
@@ -65,34 +65,41 @@ function createMessage(id) {
   return Object.assign({}, aggregateOutput[id]);
 }
 
+// Returns the arsenic values
+function getArsenicValues(region, depth) {
+  let retval = {};
```
Comment (jacekkopecky) [inline preprocessing/models/model5-estimator.js:70]: don't need retval, use return in all the branches below

File: preprocessing/models/model5-estimator.js [hunk 6]
```diff
@@ -65,34 +65,41 @@ function createMessage(id) {
   return Object.assign({}, aggregateOutput[id]);
 }
 
+// Returns the arsenic values
+function getArsenicValues(region, depth) {
```
Comment (jacekkopecky) [inline preprocessing/models/model5-estimator.js:69]: selectArsenicValues

**Review by jacekkopecky — CHANGES_REQUESTED** [review]
* clean up dropdown-data.js
* get mouza names from the mapping data, create data set with no wells but all names
* if we have any unions without mouzas, please raise it
* if we have any unions with only one mouza, add client-side code to select it automatically when the union is selected

File: preprocessing/README.md (hunk 1 above)
Comment (TxJson) [reply preprocessing/README.md]: This was fixed in commit a91f115

File: preprocessing/lib/load-data.js (hunk 2 above)
Comment (TxJson) [reply preprocessing/lib/load-data.js]: This was removed in d9bf816

File: preprocessing/lib/load-data.js (hunk 4 above)
Comment (TxJson) [reply preprocessing/lib/load-data.js]: This was corrected in d9bf816

File: preprocessing/models/model5-estimator.js (hunk 6 above)
Comment (TxJson) [reply preprocessing/models/model5-estimator.js]: This was changed in 0544d1d

File: preprocessing/models/model5-estimator.js (hunk 5 above)
Comment (TxJson) [reply preprocessing/models/model5-estimator.js]: This was changed in 0544d1d

**Review by jacekkopecky — CHANGES_REQUESTED** [review]
Please run any tests you can to check that my last commits didn't break anything.

Please clean up data/disabled/name-corrections/ – choose which ones we want to keep and say why.

Also, give a better name to data/name-corrections/*

File: data/README.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# data files with arsenic concentrations
```
[the remark this answers is not in the document]
Comment (benkyd) [reply data/README.md:1]: fixed in 3bc77ec

File: preprocessing/README.md [hunk 8]
```diff
@@ -105,3 +128,25 @@ const dropdownData = [
...
+## Tests
+
+`npm test` will run unit tests on the stats code.
+
+`sh preprocessing/tests/test-cli.sh` will run all the models against all datasets configures in `test-cli.sh`, produce output for every geographic region and every well depth we care about, and compare the output against benchmark output in `preprocessing/tests/benchmark-data`. If we aren't changing the statistical models or the underlying data sets, we should not see any differences in the generated data.
+
+### When we add/change a model
+
+A changed model should generate different outputs from the benchmark data. We can look at the differences and see that they correspond to the model changes we meant to implement. When satisfied, we can move `preprocessing/outputs/<model>` generated outputs into `benchmark-data`.
+
```
[the remark this answers is not in the document]
Comment (benkyd) [reply preprocessing/README.md:141]: fixed in 3bc77ec

File: preprocessing/cli/test-verygoodquality.js [hunk 9]
```diff
@@ -16,9 +16,11 @@ const CSV_PARSE_OPTIONS = {
...
-    : (upperQ < arsenic) ? 'underestimate'
-      : 'accurate';
+  return (lowerQ > arsenic)
+    ? 'overestimate'
+    : (upperQ < arsenic)
+        ? 'underestimate'
+        : 'accurate';
 }
 
 function runTests(allModels, div, dis, upa, uni, depth, colour, arsenic) {
```
[the remark this answers is not in the document]
Comment (benkyd) [reply preprocessing/cli/test-verygoodquality.js:26]: fixed according to suggested changes in 6d681a7

[the remark this answers is not in the document]
Comment (benkyd) [reply preprocessing/cli/test-verygoodquality.js:26]: (and 28f1c29)

File: preprocessing/geodata/centroids.js [hunk 10]
```diff
@@ -1,30 +1,115 @@
...
+}
+
+function getCentroidsIterator() {
+  if (!centroidsArray) {
+    computeCentroids();
+  }
+  return centroidsArray;
+}
+
+function annotateCentroids(divisions) {
```
[the remark this answers is not in the document]
Comment (benkyd) [reply preprocessing/geodata/centroids.js:98]: commented according to the suggested change in 762ff34

File: preprocessing/geodata/centroids.js [hunk 11]
```diff
@@ -1,30 +1,115 @@
...
+  const arr = getCentroidsIterator();
+  const lookup = getLookupCentroids();
+  for (const region of arr) {
+    try {
+      region.divisionsObj = divisions[region.div].districts[region.dis].upazilas[region.upa].unions[region.uni].mouzas[region.mou];
+      lookup[region.div].subRegions[region.dis].subRegions[region.upa].subRegions[region.uni].subRegions[region.mou].divisionsObj = region.divisionsObj;
+    } catch (e) {
+      continue;
```
[the remark this answers is not in the document]
Comment (benkyd) [reply preprocessing/geodata/centroids.js:106]: The try/catch makes sense but could probably be done in a better way, annotated in 762ff34

File: preprocessing/lib/cli-common.js [hunk 12]
```diff
@@ -23,6 +29,16 @@ const optionDefinitions = [
     alias: 'o',
     type: String,
   },
+  {
+    name: 'inputFile',
+    alias: 'i',
+    type: String,
+  },
+  {
```
[the remark this answers is not in the document]
Comment (benkyd) [reply preprocessing/lib/cli-common.js:37]: docummented in 7387851

File: preprocessing/lib/name-corrections.js [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+    // move last part of arr to arrEnd
+    arrEnd.unshift(arrCopy.pop());
+  }
+
+  // no corrections available, assume arr is correct already
+  return arr;
+}
+
+function testCorrectRegionName(expect) {
```
[the remark this answers is not in the document]
Comment (benkyd) [reply preprocessing/lib/name-corrections.js:50]: There is no code tests that use Jest like testCorrectRegionName does, also no npm-test, it all uses the sh script

File: preprocessing/geodata/centroids.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (jacekkopecky) [reply preprocessing/geodata/centroids.js:108]: please list the regions that require this catch

File: preprocessing/lib/name-corrections.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
+/*
+ * This file handles name corrections for division/district/upazila/union/mouza
+ * spelling variations. Name corrections are builtin (below) and loaded from
+ * files in /data/name-corrections
+ */
+
+const corrections = new Map();
+
+// built-in corrections for geo-data
+// todo this should live in a file of known name corrections
```
[the remark this answers is not in the document]
Comment (benkyd) [reply preprocessing/lib/name-corrections.js]: There are a few spaces in the data where the names have double spaces, it doesn't seem to break anything in my testing and name corrections are correctly applied. Proposed changes completed in 9efa0b0

File: preprocessing/lib/name-corrections.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (benkyd) [reply preprocessing/lib/name-corrections.js]: Moved to it's own file and written another mouza test case. Can be found in 9efa0b0

File: preprocessing/models/model5-preprocessor.js [hunk 15]
```diff
@@ -370,12 +412,14 @@ function main(divisions) {
 /* /////////////////////////////// */
 
 function computeWidening(divisions) {
-  forEachUnion(divisions, stratifyWells);
+  forEachMouza(divisions, stratifyWells);
 
-  computeNearbyRegions(divisions);
+  for (const region of divisions) {
+    computeNearbyRegions(region);
```
[the remark this answers is not in the document]
Comment (benkyd) [reply preprocessing/models/model5-preprocessor.js:419]: `region.nearbyRegions` array is missing in data-per-strata, and present as an empty array in model5 preprocessor, however there is nowhere in the code that appends an empty array to `region.nearbyRegions` for further processing in `computeNearbyRegions`

**Review by jacekkopecky — APPROVED** [review]

**Pull request merged**